vir/src/legacy/ast/expr_transformers.rs

```diff
@@ -63,6 +64,34 @@ impl Expr {
         }
         ExprFolderImpl { f }.fold(self)
     }
+
+    /// Visit each position.
+    pub fn visit_positions<F: FnMut(&Position)>(&self, visitor: F) {
+        struct PositionsVisitor<F: FnMut(&Position)> {
+            visitor: F,
+        }
+        impl<F: FnMut(&Position)> ExprWalker for PositionsVisitor<F> {
+            fn walk_position(&mut self, pos: &Position) {
+                (self.visitor)(pos);
+            }
+        }
+        PositionsVisitor { visitor }.walk(self);
+    }
+
+    /// Mutably visit each position.
+    pub fn visit_positions_mut<F: FnMut(&mut Position)>(&mut self, visitor: F) {
+        struct PositionsMutVisitor<G: FnMut(&mut Position)> {
+            visitor: G,
+        }
+        impl<G: FnMut(&mut Position)> ExprFolder for PositionsMutVisitor<G> {
+            fn fold_position(&mut self, mut pos: Position) -> Position {
+                (self.visitor)(&mut pos);
+                pos
+            }
+        }
+        let expr = mem::replace(self, true.into());
+        *self = PositionsMutVisitor { visitor }.fold(expr);
```

Comment:
This would be more efficient if we had a `ExprMutVisitor` trait that visits all the subexpressions via mutable references.